uaclient/api/u/pro/services/enable/v1.py

```diff
@@ -0,0 +1,167 @@
+    BoolDataValue,
+    DataObject,
+    Field,
+    StringDataValue,
+    data_list,
+)
+
+event = event_logger.get_event_logger()
+LOG = logging.getLogger(util.replace_top_level_logger_name(__name__))
+
+
+class EnableOptions(DataObject):
+    fields = [
+        Field("service", StringDataValue),
+        Field("variant", StringDataValue, False),
+        Field("enable_required_services", BoolDataValue, False),
+        Field("disable_incompatible_services", BoolDataValue, False),
+        Field("access_only", BoolDataValue, False),
+    ]
+
+    def __init__(
+        self,
+        *,
+        service: str,
+        variant: Optional[str] = None,
+        enable_required_services: bool = True,
+        disable_incompatible_services: bool = True,
+        access_only: bool = False
+    ):
+        self.service = service
+        self.variant = variant
+        self.enable_required_services = enable_required_services
+        self.disable_incompatible_services = disable_incompatible_services
+        self.access_only = access_only
+
+
+class EnableResult(DataObject, AdditionalInfo):
+    fields = [
+        Field("enabled", data_list(StringDataValue)),
+        Field("disabled", data_list(StringDataValue)),
+        Field("reboot_required", BoolDataValue),
+        Field("messages", data_list(StringDataValue)),
+    ]
+
+    def __init__(
+        self,
+        *,
+        enabled: List[str],
+        disabled: List[str],
+        reboot_required: bool,
+        messages: List[str]
+    ):
+        self.enabled = enabled
+        self.disabled = disabled
+        self.reboot_required = reboot_required
+        self.messages = messages
+
+
+def _enabled_services_names(cfg: UAConfig) -> List[str]:
+    return [s.name for s in _enabled_services(cfg).enabled_services]
+
+
+def enable(options: EnableOptions) -> EnableResult:
+    return _enable(options, UAConfig())
+
+
+def _enable(
+    options: EnableOptions,
+    cfg: UAConfig,
+) -> EnableResult:
+    event.set_event_mode(event_logger.EventLoggerMode.JSON)
+
+    if not _is_attached(cfg).is_attached:
+        raise exceptions.UnattachedError()
+
+    enabled_services_before = _enabled_services_names(cfg)
+    if options.service in enabled_services_before:
+        # nothing to do
+        return EnableResult(
+            enabled=[],
+            disabled=[],
+            reboot_required=False,
+            messages=[],
+        )
+
+    ent_cls = entitlements.entitlement_factory(
+        cfg=cfg, name=options.service, variant=options.variant or ""
+    )
+    entitlement = ent_cls(
+        cfg,
+        assume_yes=True,
+        allow_beta=True,
+        called_name=options.service,
+        access_only=options.access_only,
+    )
+
+    success = False
+    fail_reason = None
+
+    try:
+        with lock.SpinLock(
+            cfg=cfg,
+            lock_holder="u.pro.services.enable.v1",
+        ):
+            success, fail_reason = entitlement.enable(
```

Comment:
Just to double check, if we enable a service and its status is `warning`, should we notify the user somehow ?

Reply:
Talked during standup: Even if a service ends up in a warning state, it was still enabled. The enable endpoint should present warnings that happened during the enable-action, but not about the service status post-enable. To get the full status of a service, a user should be using one of the status endpoints (e.g. u.pro.status.enabled_services.v1)